### PR TITLE
[SPARK-33225][SQL] Extract AliasHelper trait

### PR DESCRIPTION
**Pull request by tanelk**
### What changes were proposed in this pull request?

Extract methods related to handling Aliases to a trait.

### Why are the changes needed?

Avoid code duplication

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing UTs cover this

### CODE REVIEW

**Comment by maropu** [comment]
NOTE: This PR comes from the comment in https://github.com/apache/spark/pull/30018#discussion_r510101973 cc: @cloud-fan @viirya 

@tanelk Thanks for the work. Have you checked all the place where `Alias`s are used in analyzer/optimizer?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AliasHelper.scala [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    // Create a map of Aliases to their values from the child projection.
+    // e.g., 'SELECT a + b AS c, d ...' produces Map(c -> Alias(a + b, c)).
+    AttributeMap(exprs.collect { case a: Alias => (a.toAttribute, a) })
+  }
+
+  /**
+   * Replace all attributes, that reference an alias, with the aliased expression
+   */
+  protected def replaceAlias(
+    expr: Expression,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AliasHelper.scala:56]: nit: 4 space indentation

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AliasHelper.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  protected def trimAliases(e: Expression): Expression = {
+    e.transformDown {
+      case Alias(child, _) => child
+      case MultiAlias(child, _) => child
+    }
+  }
+
+  protected def trimNonTopLevelAliases(e: Expression): Expression = e match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AliasHelper.scala:86]: this method will never change the input expression type, how about
`def trimNonTopLevelAliases[T <: Expression](expr: T): T`

**Comment by cloud-fan** [comment]
looks fine

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AliasHelper.scala (hunk 2 above)
Comment (tanelk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AliasHelper.scala]: Good point!

**Comment by tanelk** [comment]
> Have you checked all the place where `Alias`s are used in analyzer/optimizer?

There were some places, but I did not feel that they deserved to get a method extracted. For example:
https://github.com/tanelk/spark/blob/63abab41b40e98947a5cb340a966e657e6b03632/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L3741

**Comment by maropu** [comment]
> There were some places, but I did not feel that they deserved to get a method extracted. For example:
https://github.com/tanelk/spark/blob/63abab41b40e98947a5cb340a966e657e6b03632/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L3741

okay, thanks for the check. LGTM.

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
Thanks! Merged to master.

**Pull request closed without merging**